src/types/flow.ts

```diff
@@ -19,18 +19,27 @@ export type FlowGraph = {
   [key: string]: Node;
 };
 
-export type IndexedNode = Node & {
+export type IndexedNode = {
   id: string;
   type: ComponentType;
   parentId: string | null; // null if it is the first node
+  rootNodeId: string;
```

Comment:
It's helpful for `IndexedNode`s to pass on information about where in the `_root.edges` that node is found/nested. This means that a flat list of `IndexedNode`s can be used to find a more meaningful `StructuredNode` - both of these data structures working in tandem make it easy to perform all kinds of traversals and look-ups.